### PR TITLE
Playback

### PR DESCRIPTION
**Pull request by TriXXeD**

### CODE REVIEW

File: cha/implementaion/playback/decoding.tex [hunk 1]
```diff
@@ -0,0 +1,48 @@
+\section{Decoding MP3}
+In \cnameref{cha:initial_development} we presented our initial foundation for development, namely the Google sample implementation of a universal media player app.
+In that implementation an Android implementation, \code{android.media.MediaPlayer}, is used to control the playback of audio streams.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: Implementation Implementation Implementation

Maybe not control but rather "is in charge of playing compressed audio files"

File: cha/implementaion/playback/decoding.tex [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+The decoding from MP3 to \ac{PCM} is already handled by the base implementation from Google samples, however we have no guarantees when actions occur, as it is left up to this already implemented system and done in correlation with playing the \ac{PCM} data.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: Not just "Google samples". Android MediaPlayer is a standard Android componenet. So it's rather handled by the mediaplayer implementation from the android framework.

File: cha/implementaion/playback/decoding.tex [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+We replace the \code{android.media.MediaPlayer} class with our own implementation in order to obtain more control over the playback and decoding.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: Some design considerations are missing here. We need to touch on why we are streaming PCM and why we need that control, otherwise this is lacking support.

Comment (DelusionalLogic) [reply cha/implementaion/playback/decoding.tex]: Looking again I just think this needs to go and let the next sentence pick it up. Having it twice in a row is annoying.

File: cha/implementaion/playback/decoding.tex [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+We want more control to be able to guarantee certain characteristics about the data, such that we have the required control to properly stream the data from one device to another and perform playback commands in sync.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: That's a lot of "control" in a single sentence. Could it be restructured?

File: cha/implementaion/playback/decoding.tex [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+These guarantees are: \mnote[inline]{what are these guarantees and why do we need them? I assume i missed some}.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: These hardly seem like "guarantees" and more like features. It feels like this section isn't really leading up to anything.

File: cha/implementaion/playback/decoding.tex [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+\begin{eletterate*}
+    \item Length of a sample, such that we can properly control the audio buffers.
+    \item Exact time it takes to play a single, such that we can adjust for drift.\mnote{Explain this better Jesper, you mentioned master playback speed control?}
+    \item Audio characteristics, i.e.\ the domain knowledge introduced earlier in this chapter, such that we can manipulate the data, in particular channels to create a stereo or multichannel setup.
+\end{eletterate*}
+
+MP3 is a compressed format, meaning certain data is lost, e.g.\ we cannot confirm the length of a single sample, unlike \ac{PCM} data where we know the precise length and time it takes to play a single sample.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: I'm hesitant on this claim. "meaning certain data is lost" It's not really lost, it's just not immediately visible, and i don't know if that's always a feature of compressed formats.

File: cha/implementaion/playback/decoding.tex [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+Furthermore \ac{PCM} samples are independent of each other unlike MP3 samples, this introduces complications when working with buffers, which may have different capabilities and requirements depending on the hardware.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: I don't really know if MP3 has any concept of "samples"

I'm very unsure of your "complications with buffers". I don't know what your intention is there.

File: cha/implementaion/playback/decoding.tex [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+\bigskip
+While figuring out how to decode, we came across several mentions that decoding MP3 to \ac{PCM} using Java libraries is slow, and that for this it would be beneficial to do native development\cite{slow_java_stackoverflow}\cite{slow_java_lib}.
+As such we have chosen to develop the decoding part of our application using \ac{JNI}\@.
+
+We already had knowledge of an available open source MP3 decoding library, \textit{libmpg123}\mnote[inline]{Is this the proper way to address what we use, should i mention the mpg123 audio player?}, which we use in our implementation.
+\mnote[inline]{Please insert:issues of note or particular interest, code examples if any are interesting, important mentions related to the implementation as this is barely touched upon? Perhaps a lessons learned section/subsection as we talked about at supervisor meeting, this could include toolchain, crosscompile and perhaps our thoughts on our later discovery of mediacodec}
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/decoding.tex]: Yeah, I think this should be expanded upon.

File: cha/implementaion/playback/meta.tex [hunk 9]
```diff
@@ -0,0 +1,28 @@
+In this chapter we explore some key elements in the development of our application related to the playback of audio data.
+The chapter includes our decisions, the reasoning behind them, some of the complications we encountered and how we dealt with them.
+Before we present the implementation and what that entails, we firstly present some audio data terminology, as well as the Android sound stack, as this domain knowledge is used in our implementation.
+
+\begin{description}
+    \item [Sample Rate]\cite{sample_rate_std} \hfill \\
+    Sample rate is the number of audio samples per channel per second.
+    The rate may vary from file to file with $44,100\ Hz$ being the de facto and most supported sampling rate.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/meta.tex:8]: That's a dubious claim 48000 is actually also supported in a lot of places. I'd rather say that 44100 is a common CD sampling rate.

File: cha/implementaion/playback/meta.tex [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+    \item [Bit Depth]\cite{bit_depth} \hfill \\
+    Bit depth is largely dependant on the data encoding used, and if relevant also described alongside sample rate, e.g.\ a data rate of 44.1 kHz/16 translates to a sample rate of 44,100 and a bit depth of 16.
+    Bit depth is the number of bits used per sample.
+    In our project we work with MP3 and PCM encoding, and as such we are only concerned with these two formats.
+    MP3 does not encode any bit depth in its data, as such the bit depth for the \ac{PCM} data depends on the decoder settings.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/meta.tex:13]: I don't actually think this makes any sense. MP3 stores data in the frequency domain in a huffman coded model. Talking about sampling rate and bitdepth doesn't make any sense in a format like that. What is true is that the "resampling" of an MP3 file can be done at any bitdepth. That means that mpg123 can decode the frequency domain mp3 data into 16bit 8bit or float time domain data.

It's pretty complex and i don't understand all of it. This paper has a nice diagram: https://www.mp3-tech.org/programmer/docs/mp3_theory.pdf

File: cha/implementaion/playback/meta.tex [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    As for PCM, this does contain bit depth data, and once again this may differ for each file with 16 and 24 bit both being common with 16-bit being the standard for CD, both Blu-ray and DVD also supports 24-bit.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/meta.tex:14]: It doesn't contain it, but it has a well defined bit depth that is uniform for every sample in the dataset. You can't somehow get it from the PCM data, you need to pass it along beside the actual data. Otherwise it's impossible to decode into soundwaves.

File: cha/implementaion/playback/meta.tex [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+    \item [Channels]\cite{channels} \hfill \\
+    Channels refer to the number of audio tracks a file contains, this can be divided into three classes: Monaural Sound (Mono), Stereophonic Sound (Stereo) and Surround/Multichannel Sound.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/meta.tex:16]: Why is stereo different from Multichannel? How is mono different from those? It's true that i can be divided into those classes, but that seems arbitrary.

File: cha/implementaion/playback/meta.tex [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+    Mono contains a single audio track, stereo contains two audio tracks, and anything beyond that can be considered multichannel.
+    Stereo is the most common of the aforementioned classes.
+    The number of channels also affects the bit rate.
+    \item [Bit Rate]\cite{bit_rate} \hfill \\
+    Bit rate, in respect to media, refers to the number of bits used per unit of playback time.
+    Bit rate depends on both sample rate, encoding in the sense of bit depth, and number of channels in that it is calculated by $sample rate \times bit depth \times channels$.
+    \item [Frame Size]\cite{audioformat} \hfill \\
+    A frame is defined as all the data that belongs to one sampling interval, for PCM data this also means that the frame rate is equal to the sample rate.
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/meta.tex:24]: That's not true for MP3, where a frame is an arbitrarily sized amount of data that they group together for whatever reason.

File: cha/implementaion/playback/playing.tex [hunk 14]
```diff
@@ -0,0 +1,50 @@
+\section{Playing \ac{PCM} Data}
+With the data decoded we also create a replacement for the mediaplayer.\mnote[inline]{why btw? could we not simply have used the one we alrdy had?}
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/playing.tex:2]: The decoding is part of the default mediaplayer, and therefore it's part of the replacement.

File: cha/implementaion/playback/playing.tex [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+In preparation for the next part of development, enabling devices to connect to one another and create a multi-device system, we treat the data as an incoming stream rather than local data.
+To handle the \ac{PCM} data stream we use the android class \code{AudioTrack} which allows for streaming of PCM audio buffers to an audio sink.
+
+\mnote[inline]{the following is largely related to how AudioTrack works with the hardware, it would greatly benefit from some implementation related issues and observations}
+The \code{AudioTrack} class does impose a couple of requirements on how we handle the data stream, particularly impacting or playback commands.
+First off \code{AudioTrack} requires that we constantly write to it, if this does not happen it will pause in order to release system resources, this can never happen as it would introduce wake up lag upon the next write action.
+Secondly \code{AudioTrack} works with an internal buffer, this buffer has to be completely filled before it can be flushed to the audio sink, which means the data is send in small buffers at a time.\cite{audiotrack}
```
Comment (DelusionalLogic) [inline cha/implementaion/playback/playing.tex:9]: "which means the data is send in small buffers at a time" What how is that related to the rest of the sentence?

**Review by DelusionalLogic — CHANGES_REQUESTED** [review]
I might seem a bit down on this, and that's because i am.

I don't feel like it does a good job at describing what we have in the app today, or the challenges we happened upon. It's also really lacking a good argument for why we are even doing this. What goal does is help it us achieve? If i was reading this for the first time I would wonder if it was just self masturbatory programming, because we didn't have anything better to do.

It's in a design part of the paper, but doesn't include any information about design, just some vague definitions of various file formats and android framework classes. If i was reading this for the first time I would wonder if it was just self masturbatory programming, because we didn't have anything better to do.

File: cha/implementaion/playback/decoding.tex (hunk 6 above)
Comment (TriXXeD) [reply cha/implementaion/playback/decoding.tex]: True, maybe say lossy instead, as it is a lossy compression

File: cha/implementaion/playback/meta.tex (hunk 9 above)
Comment (TriXXeD) [reply cha/implementaion/playback/meta.tex:8]: It's not something i claim, it is taken from the source

File: cha/implementaion/playback/meta.tex (hunk 12 above)
Comment (TriXXeD) [reply cha/implementaion/playback/meta.tex:16]: I grouped it like this because Mono and Stereo are relevant for us, thus i grouped the remaining, splitting it into quad, 5.1, 6.1, 7.1 would just be a worthless expansion imo.

File: cha/implementaion/playback/playing.tex (hunk 15 above)
Comment (TriXXeD) [reply cha/implementaion/playback/playing.tex:9]: It's essentially explaining the rest of the sentence, its related but completely superfluous
